src/functions.jl

```diff
@@ -146,6 +146,78 @@ function Base.append!(A::NamedDimsArray{L,T,1}, B::AbstractVector) where {L,T}
     return NamedDimsArray{newL}(data)
 end
 
+################################################
+# cat, vcat, hcat
+
+function Base.cat(a::NamedDimsArray{L}; dims) where L
+    newL = expand_dimnames(L, dims)
+    numerical_dims = dim(newL, dims)
+    data = Base.cat(parent(a); dims=numerical_dims)
+    newL = unify_names_longest(newL, dimnames(data))
+    return NamedDimsArray{newL}(data)
+end
+
+function Base.cat(a::NamedDimsArray{L}, b::AbstractArray; dims) where L
+    newL = expand_dimnames(L, dims)
+    numerical_dims = dim(newL, dims)
+    data = Base.cat(parent(a), b; dims=numerical_dims)
+    newL = unify_names_longest(newL, dimnames(data)) # when dims=3 for two 2d arrays
```

Comment:
It might be good to check closely if `data` is type-infering right and that calling `dimnames` on data returns a constant result.
We should have basically all the info we need at compile time to workout what the final names will be.
So this should be fully type stable, but it seems not to be to me, so we might need a different way to find the names than usign `dimnames` on `cat`